// rollup-plugin-babel/dist/rollup-plugin-babel.cjs.js
'use strict';

var path = require('path');
var _babel_core = require('@babel/core');
var rollupPluginutils = require('rollup-pluginutils');

var INLINE = {};
var RUNTIME = {};
var EXTERNAL = {};

var HELPERS = 'rollupPluginBabelHelpers';

function fallbackClassTransform () {
	return {
		visitor: {
			ClassDeclaration: function ClassDeclaration (path$$1, state) {
				path$$1.replaceWith(path$$1, state.file.addHelper('inherits'));
			}
		}
	};
}

var preflightCheckResults = {};

function preflightCheck ( options, dir ) {
	if ( !preflightCheckResults[ dir ] ) {
		var helpers;

		options = Object.assign( {}, options );
		delete options.only;
		delete options.ignore;

		options.filename = path.join( dir, 'x.js' );

		var inputCode = 'class Foo extends Bar {};\nexport default Foo;';
		var check = _babel_core.transform( inputCode, options ).code;

		if ( ~check.indexOf('class ') ) {
			options.plugins = (options.plugins || []).concat( fallbackClassTransform );
			check = _babel_core.transform( inputCode, options ).code;
		}

		if ( ~check.indexOf( 'import _inherits' ) ) { helpers = RUNTIME; }
		else if ( ~check.indexOf( 'function _inherits' ) ) { helpers = INLINE; }
		else if ( ~check.indexOf( 'babelHelpers' ) ) { helpers = EXTERNAL; }
		else {
			throw new Error( 'An unexpected situation arose. Please raise an issue at https://github.com/rollup/rollup-plugin-babel/issues. Thanks!' );
		}

		if ( !~check.indexOf( 'export default' ) && !~check.indexOf( 'export default Foo' ) ) { throw new Error( 'It looks like your Babel configuration specifies a module transformer. Please disable it. See https://github.com/rollup/rollup-plugin-babel#configuring-babel for more information' ); }

		preflightCheckResults[ dir ] = helpers;
	}

	return preflightCheckResults[ dir ];
}

// that is basically babel-plugin-external-helpers
function importHelperPlugin (ref) {
	var t = ref.types;

	return {
		pre: function pre (file) {
			file.set('helpersNamespace', t.identifier(HELPERS));
		}
	};
}

var warned = {};
function warnOnce ( warn, msg ) {
	if ( warned[ msg ] ) { return; }
	warned[ msg ] = true;
	warn( msg );
}

function babel ( options ) {
	options = Object.assign( {}, options || {} );

	var filter = rollupPluginutils.createFilter( options.include, options.exclude );
	delete options.include;
	delete options.exclude;

	if ( options.sourceMap !== false ) { options.sourceMaps = true; }
	if ( options.sourceMaps !== false ) { options.sourceMaps = true; }
	delete options.sourceMap;

	var runtimeHelpers = options.runtimeHelpers;
	delete options.runtimeHelpers;

	var externalHelpers;
	if ( options.externalHelpers ) { externalHelpers = true; }
	delete options.externalHelpers;

	var externalHelpersWhitelist = null;
	if ( options.externalHelpersWhitelist ) { externalHelpersWhitelist = options.externalHelpersWhitelist; }
	delete options.externalHelpersWhitelist;

	var warn = function (msg) { return console.warn(msg); }; // eslint-disable-line no-console

	return {
		name: 'babel',

		options: function options$1 ( options ) {
			warn = options.onwarn || warn;
		},

		resolveId: function resolveId ( id ) {
			if ( id === HELPERS ) { return id; }
		},

		load: function load ( id ) {
			if ( id !== HELPERS ) {
				return;
			}

			return _babel_core.buildExternalHelpers( externalHelpersWhitelist, 'module' );
		},

		transform: function transform$1 ( code, id ) {
			if ( !filter( id ) ) { return null; }
			if ( id === HELPERS ) { return null; }

			var helpers = preflightCheck( options, path.dirname( id ) );

			if ( helpers === EXTERNAL && !externalHelpers ) {
				warnOnce( warn, 'Using "external-helpers" plugin with rollup is deprecated. "rollup-plugin-babel" knows at its own how to deduplicate your babel helpers.' );
			} else if ( helpers === RUNTIME && !runtimeHelpers ) {
				throw new Error( 'Runtime helpers are not enabled. Either exclude the transform-runtime Babel plugin or pass the `runtimeHelpers: true` option. See https://github.com/rollup/rollup-plugin-babel#configuring-babel for more information' );
			}

			var localOpts = Object.assign({ filename: id }, options);

			if ( helpers !== RUNTIME ) {
				localOpts = Object.assign({}, localOpts, { plugins: (localOpts.plugins || []).concat(importHelperPlugin) });
			}

			var transformed = _babel_core.transform( code, localOpts );

			if ( helpers !== RUNTIME ) {
				transformed.code += "\n\nimport * as " + HELPERS + " from '" + HELPERS + "';";
			}

			return {
				code: transformed.code,
				map: transformed.map
			};
		}
	};
}

module.exports = babel;
