// babel-cli/lib/babel/dir.js
"use strict";

exports.__esModule = true;

exports.default = function (commander, filenames, opts) {
  function write(src, relative, base) {
    if (!util.isCompilableExtension(relative, commander.extensions)) {
      return false;
    }

    relative = util.adjustRelative(relative, commander.keepFileExtension);
    var dest = getDest(commander, relative, base);
    var data = util.compile(src, (0, _defaults2.default)({
      sourceFileName: (0, _slash2.default)(_path2.default.relative(dest + "/..", src)),
      sourceMapTarget: _path2.default.basename(relative)
    }, opts));
    if (!data) return false;

    if (data.map && commander.sourceMaps && commander.sourceMaps !== "inline") {
      var mapLoc = dest + ".map";
      data.code = util.addSourceMappingUrl(data.code, mapLoc);
      (0, _outputFileSync2.default)(mapLoc, JSON.stringify(data.map));
    }

    (0, _outputFileSync2.default)(dest, data.code);
    util.chmod(src, dest);
    util.log(src + " -> " + dest);
    return true;
  }

  function getDest(commander, filename, base) {
    if (commander.relative) return _path2.default.join(base, commander.outDir, filename);
    return _path2.default.join(commander.outDir, filename);
  }

  function handleFile(src, filename, base) {
    var didWrite = write(src, filename, base);

    if (!didWrite && commander.copyFiles) {
      var dest = getDest(commander, filename, base);
      (0, _outputFileSync2.default)(dest, _fs2.default.readFileSync(src));
      util.chmod(src, dest);
    }
  }

  function handle(filename) {
    if (!_fs2.default.existsSync(filename)) return;

    var stat = _fs2.default.statSync(filename);

    if (stat.isDirectory(filename)) {
      var dirname = filename;

      if (commander.deleteDirOnStart) {
        util.deleteDir(commander.outDir);
      }

      util.readdir(dirname).forEach(function (filename) {
        var src = _path2.default.join(dirname, filename);

        handleFile(src, filename, dirname);
      });
    } else {
      write(filename, _path2.default.basename(filename), _path2.default.dirname(filename));
    }
  }

  if (!commander.skipInitialBuild) {
    filenames.forEach(handle);
  }

  if (commander.watch) {
    var chokidar = util.requireChokidar();
    filenames.forEach(function (dirname) {
      var watcher = chokidar.watch(dirname, {
        persistent: true,
        ignoreInitial: true,
        awaitWriteFinish: {
          stabilityThreshold: 50,
          pollInterval: 10
        }
      });
      ["add", "change"].forEach(function (type) {
        watcher.on(type, function (filename) {
          var relative = _path2.default.relative(dirname, filename) || filename;

          try {
            handleFile(filename, relative);
          } catch (err) {
            console.error(err.stack);
          }
        });
      });
    });
  }
};

var _defaults = require("lodash/defaults");

var _defaults2 = _interopRequireDefault(_defaults);

var _outputFileSync = require("output-file-sync");

var _outputFileSync2 = _interopRequireDefault(_outputFileSync);

var _slash = require("slash");

var _slash2 = _interopRequireDefault(_slash);

var _path = require("path");

var _path2 = _interopRequireDefault(_path);

var _fs = require("fs");

var _fs2 = _interopRequireDefault(_fs);

var _util = require("./util");

var util = _interopRequireWildcard(_util);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
