// babel-cli/lib/babel/file.js
"use strict";

exports.__esModule = true;

exports.default = function (commander, filenames, opts) {
  if (commander.sourceMaps === "inline") {
    opts.sourceMaps = true;
  }

  var results = [];

  var buildResult = function buildResult() {
    var map = new _sourceMap2.default.SourceMapGenerator({
      file: _path2.default.basename(commander.outFile || "") || "stdout",
      sourceRoot: opts.sourceRoot
    });
    var code = "";
    var offset = 0;
    results.forEach(function (result) {
      code += result.code + "\n";

      if (result.map) {
        var consumer = new _sourceMap2.default.SourceMapConsumer(result.map);
        var sources = new Set();
        consumer.eachMapping(function (mapping) {
          if (mapping.source != null) sources.add(mapping.source);
          map.addMapping({
            generated: {
              line: mapping.generatedLine + offset,
              column: mapping.generatedColumn
            },
            source: mapping.source,
            original: mapping.source == null ? null : {
              line: mapping.originalLine,
              column: mapping.originalColumn
            }
          });
        });
        sources.forEach(function (source) {
          var content = consumer.sourceContentFor(source, true);

          if (content !== null) {
            map.setSourceContent(source, content);
          }
        });
        offset = code.split("\n").length - 1;
      }
    });

    if (commander.sourceMaps === "inline" || !commander.outFile && commander.sourceMaps) {
      code += "\n" + _convertSourceMap2.default.fromObject(map).toComment();
    }

    return {
      map: map,
      code: code
    };
  };

  var output = function output() {
    var result = buildResult();

    if (commander.outFile) {
      if (commander.sourceMaps && commander.sourceMaps !== "inline") {
        var mapLoc = commander.outFile + ".map";
        result.code = util.addSourceMappingUrl(result.code, mapLoc);

        _fs2.default.writeFileSync(mapLoc, JSON.stringify(result.map));
      }

      _fs2.default.writeFileSync(commander.outFile, result.code);
    } else {
      process.stdout.write(result.code + "\n");
    }
  };

  var stdin = function stdin() {
    var code = "";
    process.stdin.setEncoding("utf8");
    process.stdin.on("readable", function () {
      var chunk = process.stdin.read();
      if (chunk !== null) code += chunk;
    });
    process.stdin.on("end", function () {
      results.push(util.transform(commander.filename, code, (0, _defaults2.default)({
        sourceFileName: "stdin"
      }, opts)));
      output();
    });
  };

  var walk = function walk() {
    var _filenames = [];
    results = [];
    filenames.forEach(function (filename) {
      if (!_fs2.default.existsSync(filename)) return;

      var stat = _fs2.default.statSync(filename);

      if (stat.isDirectory()) {
        var dirname = filename;
        util.readdirFilter(filename).forEach(function (filename) {
          _filenames.push(_path2.default.join(dirname, filename));
        });
      } else {
        _filenames.push(filename);
      }
    });

    _filenames.forEach(function (filename) {
      var sourceFilename = filename;

      if (commander.outFile) {
        sourceFilename = _path2.default.relative(_path2.default.dirname(commander.outFile), sourceFilename);
      }

      sourceFilename = (0, _slash2.default)(sourceFilename);
      var data = util.compile(filename, (0, _defaults2.default)({
        sourceFileName: sourceFilename
      }, opts));
      if (!data) return;
      results.push(data);
    });

    output();
  };

  var files = function files() {
    if (!commander.skipInitialBuild) {
      walk();
    }

    if (commander.watch) {
      var chokidar = util.requireChokidar();
      chokidar.watch(filenames, {
        persistent: true,
        ignoreInitial: true,
        awaitWriteFinish: {
          stabilityThreshold: 50,
          pollInterval: 10
        }
      }).on("all", function (type, filename) {
        if (!util.isCompilableExtension(filename, commander.extensions)) {
          return;
        }

        if (type === "add" || type === "change") {
          util.log(type + " " + filename);

          try {
            walk();
          } catch (err) {
            console.error(err.stack);
          }
        }
      });
    }
  };

  if (filenames.length) {
    files();
  } else {
    stdin();
  }
};

var _convertSourceMap = require("convert-source-map");

var _convertSourceMap2 = _interopRequireDefault(_convertSourceMap);

var _defaults = require("lodash/defaults");

var _defaults2 = _interopRequireDefault(_defaults);

var _sourceMap = require("source-map");

var _sourceMap2 = _interopRequireDefault(_sourceMap);

var _slash = require("slash");

var _slash2 = _interopRequireDefault(_slash);

var _path = require("path");

var _path2 = _interopRequireDefault(_path);

var _fs = require("fs");

var _fs2 = _interopRequireDefault(_fs);

var _util = require("./util");

var util = _interopRequireWildcard(_util);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
